src/engine/Service.cpp

```diff
@@ -192,24 +198,26 @@ std::optional<std::string> Service::getSiblingValuesClause() const {
     if (it == siblingVars.end()) {
       continue;
     }
-    vars += it->first.name() + " ";
+    absl::StrAppend(&vars, it->first.name(), " ");
     commonColumnIndices.push_back(it->second.columnIndex_);
   }
   vars.back() = ')';
 
+  checkCancellation();
+
   ad_utility::HashSet<std::string> rowSet;
 
   std::string values = " { ";
   for (size_t rowIndex = 0; rowIndex < siblingResult->size(); ++rowIndex) {
     std::string row = "(";
-    for (size_t i = 0; i < commonColumnIndices.size(); ++i) {
+    for (const auto& columnIdx : commonColumnIndices) {
```

Comment:
Inide the for each row loop (the outer one) should probably be one cancellation check per iterationl